[Merged by Bors] - Implement expected withdrawals endpoint

eserilev opened the pull request:
## Issue Addressed

[#4029](https://github.com/sigp/lighthouse/issues/4029)

## Proposed Changes

implement expected_withdrawals HTTP API per the spec 

https://github.com/ethereum/beacon-APIs/pull/304

## Additional Info

eserilev committed:
create get expected withdrawals endpoint

eserilev committed:
added response object

eserilev committed:
added some unit tests

eserilev committed:
unit test refactor

eserilev committed:
tests

eserilev committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into implement-expected-withdrawls-endpoint

eserilev committed:
cleanup

eserilev committed:
cleanup

eserilev committed:
fmt

jimmygchen commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -1933,6 +1933,51 @@ pub fn serve<T: BeaconChainTypes>(
+        .and(warp::path("expected_withdrawals"))
+        .and(warp::path::end())
+        .and_then(|chain: Arc<BeaconChain<T>>, state_id: StateId| {
+            blocking_json_task(move || {
```
Looks like this endpoint also need to support SSZ responses:
https://github.com/ethereum/beacon-APIs/blob/a62aa810fb283873de0d8025405f7a8ce6991f51/apis/builder/states/expected_withdrawals.yaml#L37-L39

jimmygchen commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -1933,6 +1933,51 @@ pub fn serve<T: BeaconChainTypes>(
+                            finalized: finalized,
+                        },
+                    )),
+                    //.map(|resp| )),
```
I think this commented line can be removed.

jimmygchen commented on common/eth2/src/lib.rs:
```diff
@@ -1046,6 +1046,23 @@ impl BeaconNodeHttpClient {
+    pub async fn get_expected_withdrawals<T:EthSpec>(
+        &self,
+        state_id: &StateId,
+    ) -> Result<GenericResponse<NextWithdrawalsResponse<VariableList<Withdrawal,T>>>, Error> {
```
Looks like this isn't formatted with `rustfmt` which you can fix with `cargo fmt --all`.

jimmygchen commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -1933,6 +1933,51 @@ pub fn serve<T: BeaconChainTypes>(
+        .and(warp::path("states"))
+        .and(chain_filter.clone());
+
+    // GET builder/states/{state_id}/expected_withdrawals
```
This endpoint also accepts an optional `proposal_slot` query string.

https://ethereum.github.io/beacon-APIs/?urls.primaryName=dev#/Builder/getNextWithdrawals

jimmygchen commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -1933,6 +1933,51 @@ pub fn serve<T: BeaconChainTypes>(
+                    .canonical_head
+                    .cached_head()
+                    .forkchoice_update_parameters();
+                match chain.get_expected_withdrawals(&forkchoice_update_parameters, state.slot()) {
```
I think the 2nd parameter here would be the proposer slot, which would be `state.slot() + 1`?

jimmygchen commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -1933,6 +1933,51 @@ pub fn serve<T: BeaconChainTypes>(
+                    Ok(withdrawals) => Ok(api_types::GenericResponse::from(
+                        api_types::NextWithdrawalsResponse {
+                            data: withdrawals,
+                            execution_optimistic: execution_optimistic,
+                            finalized: finalized,
```
Clippy lints will error here - the redundant field names should be removed. You can run `make lint` to run lint locally.

```suggestion
                            execution_optimistic,
                            finalized,
```

jimmygchen commented on beacon_node/http_api/tests/tests.rs:
```diff
@@ -4892,3 +4936,43 @@ async fn optimistic_responses() {
+#[tokio::test(flavor = "multi_thread", worker_threads = 2)]
+async fn expected_withdrawals_invalid_state() {
+    let mut config = ApiTesterConfig {
+        builder_threshold: Some(0),
```
Not sure if we need `builder_threshold` here, can probably go with `ApiTesterConfig::default()`? (same for the test below)

jimmygchen commented:
@eserilev thanks for this! I've added a few comments but overall looks great to me 👍

eserilev committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into implement-expected-withdrawls-endpoint

eserilev committed:
cargo fmt

eserilev committed:
add optional query param

eserilev committed:
adding ssz encoding

eserilev committed:
fmt

eserilev committed:
cleanup test

eserilev commented:
thanks for the feedback @jimmygchen , this should be ready for another review

eserilev committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into implement-expected-withdrawls-endpoint

jimmygchen commented:
Thanks, will take a look at this tomorrow 😊

jimmygchen commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -1933,6 +1933,77 @@ pub fn serve<T: BeaconChainTypes>(
+    let get_expected_withdrawals = builder_states_path
+        .clone()
+        .and(warp::path::param::<StateId>())
+        .and(warp::query::<api_types::ExpectedWithdrawalsQuery>())
```
I think `query` generally gets added after `path` params - I'm not 100% sure whether the order matters with `warp`, this might still work but I think would be a bit more consistent and readable to have it at the end.

jimmygchen commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -1933,6 +1933,77 @@ pub fn serve<T: BeaconChainTypes>(
+             query: api_types::ExpectedWithdrawalsQuery,
+             accept_header: Option<api_types::Accept>| {
+                blocking_response_task(move || {
+                    let proposal_slot = query.proposal_slot.unwrap_or(Slot::new(1));
```
The default `proposal_slot` should be the slot after the pre state if it's  not specified; when it's specified the `proposal_slot` should just be `proposal_slot`.

I think what you have works for the default case where `proposer_slot` is not specified (slot = `state.slot() + 1`), but if the current `state.slot` is 100503, and caller specifies a `proposer_slot` of 100504, it ends up adding both numbers up (`state.slot() + proposal_slot`), and this gives you the incorrect proposal slot (100503 + 100504 = 201007).

I think you may want something like:

```rust
let proposal_slot = query.proposal_slot.unwrap_or(state.slot() + 1);
```

jimmygchen replied:
I think we should also validate that the `proposal_slot` is greater than `state.slot`, as we are computing the expected withdrawals based on the pre-state and slot number, which must be a slot after the state.

jimmygchen replied:
I wonder how far ahead of the prestate we should allow `proposal_slot` to be - if user provides a far future slot, we should probably return an error rather than computing it - although I haven't seen any specific number in the spec.

eserilev replied:
> I wonder how far ahead of the prestate we should allow `proposal_slot` to be - if user provides a far future slot, we should probably return an error rather than computing it - although I haven't seen any specific number in the spec.

thats a good point. looking at the prysm [implementation](https://github.com/prysmaticlabs/prysm/pull/12519/files) of this endpoint they calculated a look ahead limit using MaxSeedLookahead

`lookAheadLimit := uint64(params.BeaconConfig().SlotsPerEpoch.Mul(uint64(params.BeaconConfig().MaxSeedLookahead)))`

and then returned an error if `proposal slot >= state.slot + lookAheadlimit`

I will try to do something similar to this

jimmygchen commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -1933,6 +1933,77 @@ pub fn serve<T: BeaconChainTypes>(
+             accept_header: Option<api_types::Accept>| {
+                blocking_response_task(move || {
+                    let proposal_slot = query.proposal_slot.unwrap_or(Slot::new(1));
+                    let (state, execution_optimistic, finalized) = match state_id.state(&chain) {
```
I think we can safely `?` the `state_id.state` as the "state not found" error is already being handled and returned from the `state_id.state` function here:

https://github.com/sigp/lighthouse/blob/cb472d21df859efd0870fb43956fe0fcd3aea905/beacon_node/http_api/src/state_id.rs#L178-L188

jimmygchen commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -1933,6 +1933,77 @@ pub fn serve<T: BeaconChainTypes>(
+                        Ok(withdrawals) => withdrawals,
+                        Err(_) => {
+                            return Err(warp_utils::reject::custom_bad_request(
+                                "The specified state is not a capella state.".to_string(),
```
I think this error may be incorrect as an error from `chain.get_expected_withdrawals` does not indicate that the state isn't a capella state. It seems like errors from this function indicate an unexpected condition and should probably return 500. 

Although I think we should validate that the `proposal_slot` is a slot after the capella fork, and return an error if the `proposal_slot` is pre-capella.

jimmygchen commented on beacon_node/http_api/tests/tests.rs:
```diff
@@ -4152,6 +4152,53 @@ impl ApiTester {
+        let slot = self.chain.slot().unwrap();
+        let state_id = CoreStateId::Slot(slot);
+
+        let result = self.client.get_expected_withdrawals::<E>(&state_id).await;
```
I think calling `unwrap` on this result achieves the same thing without the below `match` statement, given we're not making any assertions here - although I wonder if it's possible to verify the withdrawals in the response?

eserilev replied:
I decided to update the test to calculate the expected withdrawal, and then make a request to the client to compare the expected withdrawal calculated vs the withdrawal response that the client returns

jimmygchen commented:
@eserilev this is looking good, thank you! I've left a few more comments.  I'm not super familiar with capella / withdrawal code, would be helpful to get a 2nd review from @ethDreamer or @michaelsproul 🙏

paulhauner commented:
I've flagged this as v4.3.0 since it's an isolated change and we've had some interest in getting it in for v4.3.0.

That being said, we shouldn't delay v4.3.0 for this since Gnosis needs a new release rather urgently. If we start to delay let's bump it to v4.3.1.

michaelsproul commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -1933,6 +1933,77 @@ pub fn serve<T: BeaconChainTypes>(
+                            ))
+                        }
+                    };
+                    let forkchoice_update_parameters = chain
+                        .canonical_head
+                        .cached_head()
+                        .forkchoice_update_parameters();
+
+                    let withdrawals = match chain.get_expected_withdrawals(
+                        &forkchoice_update_parameters,
+                        state.slot() + proposal_slot,
+                    ) {
```
We shouldn't use `chain.get_expected_withdrawals` here, because it requires the `forkchoice_update_parameters`, which currently are being read from the head state rather than the state that was requested.

Instead, we should use the state that was loaded, and call `get_expected_withdrawals(&state, &chain.spec)`. Before computing the withdrawals this state should be _advanced_ to the epoch of the proposal slot using some logic like this:

```rust
let proposal_epoch = proposal_slot.epoch(E::slots_per_epoch());
if proposal_epoch != state.current_epoch() {
    partial_state_advance(
        &mut state,
        None,
        proposal_slot,
        &self.spec,
    )?;
}
```

eserilev committed:
changes based on feedback

eserilev committed:
add look ahead limit

eserilev committed:
merge conflict

eserilev committed:
better testing

eserilev commented:
this should be ready for another look. thank you!

jimmygchen commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -2018,6 +2022,104 @@ pub fn serve<T: BeaconChainTypes>(
+                blocking_response_task(move || {
+                    let (mut state, execution_optimistic, finalized) = state_id.state(&chain)?;
+                    let proposal_slot = query.proposal_slot.unwrap_or(state.slot() + 1);
+                    if proposal_slot < state.slot() {
```
Proposal slot has to be greater than the pre-state slot:

```suggestion
                    if proposal_slot <= state.slot() {
```

jimmygchen commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -2018,6 +2022,104 @@ pub fn serve<T: BeaconChainTypes>(
+                    let fork = chain.spec.fork_name_at_slot::<T::EthSpec>(proposal_slot);
+                    if let ForkName::Base | ForkName::Altair | ForkName::Merge = fork {
+                        return Err(warp_utils::reject::custom_bad_request(
+                            "the specified state is not a capella state.".to_string(),
```
This endpoint works for post-capella slots (including `deneb` and so on). Probably a more accurate error message would be:

```suggestion
                            "the specified proposal slot is a pre-capella slot.".to_string(),
```

michaelsproul commented:
Dropping this to v4.3.1 as there's no longer any urgency. The demand for this API from users is better satisfied by loading withdrawals from blocks (without any expensive state lookups)

jimmygchen commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -2018,6 +2022,104 @@ pub fn serve<T: BeaconChainTypes>(
+                        .max_seed_lookahead
+                        .into_inner()
+                        .safe_mul(T::EthSpec::slots_per_epoch())
+                        .unwrap();
```
We generally avoid using `unwrap` or functions that could cause panic in non testing code, perhaps we could use `warp_utils::reject::arith_error` here?

It might be worth moving the endpoint logic to a separate file & function now this has grown a bit, I think it would also make error handling a bit easier because you could make use of automatic error propagation `?` with `map_err(error_type)?`.

jimmygchen commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -2018,6 +2022,104 @@ pub fn serve<T: BeaconChainTypes>(
+
+                    let look_ahead_limit = chain
+                        .spec
+                        .max_seed_lookahead
```
I think the value of `4` epochs seems reasonable but I'm not sure if [this constant](https://eth2book.info/capella/annotated-spec/#max_seed_lookahead) was designed for this purpose, perhaps introducing a constant with a small number of epochs would be fine?

jimmygchen commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -2018,6 +2022,104 @@ pub fn serve<T: BeaconChainTypes>(
+                    }
+
+                    let proposal_epoch = proposal_slot.epoch(T::EthSpec::slots_per_epoch());
+                    let (state_root, _, _) = state_id.root(&chain).unwrap();
```
We should avoid `unwrap` here and return an error instead if state is not found.

eserilev committed:
add arth error, and some cleanup

eserilev committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into implement-expected-withdrawls-endpoint

eserilev committed:
refacor based on feedback

eserilev committed:
small revert

eserilev committed:
small revert

eserilev committed:
cleanup

eserilev committed:
look ahead constant

eserilev commented:
apologies for all the back and forth on this one & thank you for the multiple reviews. I refactored based on your feedback and moved most of the logic to its own file. It should be ready for another review, thank you!

jimmygchen commented:
> apologies for all the back and forth on this one & thank you for the multiple reviews. I refactored based on your feedback and moved most of the logic to its own file. It should be ready for another review, thank you!

Thank _you_ for implementing this 🙏 I'll take a look some time this week when I can!

jimmygchen reviewed (approved):
Looks great to me, thank you @eserilev :)  Would be great to get a quick sanity check from @michaelsproul and we should be good to merge!

eserilev committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into implement-expected-withdrawls-endpoint

eserilev commented:
merged latest unstable

eserilev committed:
resolve merge conflict

michaelsproul commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -2105,6 +2107,56 @@ pub fn serve<T: BeaconChainTypes>(
+             state_id: StateId,
+             query: api_types::ExpectedWithdrawalsQuery,
+             accept_header: Option<api_types::Accept>| {
+                blocking_response_task(move || {
```
We need to refactor this to use the `task_spawner` that was added in #4462. The surrounding endpoints should give an idea of how to do this: add the `task_spawner` filter, then call its `blocking_response_task` instead.

michaelsproul commented on common/eth2/src/types.rs:
```diff
@@ -1267,6 +1272,14 @@ pub struct ForkChoiceNode {
 
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
+#[serde(bound = "T: Serialize + serde::de::DeserializeOwned")]
+pub struct NextWithdrawalsResponse<T: Serialize + serde::de::DeserializeOwned> {
```
Could use `ExecutionOptimisticFinalizedResponse` instead of adding this new type, but I don't mind too much either way.

michaelsproul reviewed (changes requested):
Looking good! Just a couple more changes then I think we should be good to merge for v4.4.0

eserilev committed:
remove unecessary struct

eserilev committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into implement-expected-withdrawls-endpoint

eserilev committed:
add task spawner

eserilev commented:
thanks! should be ready for another review

jimmygchen commented:
@eserilev changes look great! 

I've triggered another CI run and looks like`cargo fmt` failed, you can fix this with `cargo fmt --all`.

eserilev committed:
fmt

eserilev commented:
whoops, that should be resolved now

jimmygchen reviewed (approved):
Looks great, thank you! 😃 

The cargo audit CI failure is not related to this PR, and will be addressed in #4651.

michaelsproul commented:
bors r+

michaelsproul commented:
bors r-
bors r+
